content/posts/introducing-krator.md

```diff
@@ -0,0 +1,359 @@
+considered using macros for implementing `State`, but have not been satisfied
+with Rust's error messages for errors originating from within these macros. If
+you have any experience with macros and think you can come up with something
+nice, please let us know!
+
+The `next` method is the main code block for each state, sometimes referred to
+as the state handler. This method has a number of important arguments:
+
+- It takes ownership of `self`.
+- An `Arc<RwLock<_>>` to access shared state. For more information on this API,
+  see the
+  [Tokio documentation](https://docs.rs/tokio/1.0.2/tokio/sync/struct.RwLock.html).
+- A mutable reference to the object state.
+- A `Manifest<_>` which acts as a
+  [reflector](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes).
+  This type implements the `Stream` API for detecting changes to the object
+  manifest within a state handler, or you can simply call `latest` to obtain
+  a clone of the current object manifest.
+  
+
+```rust
+#[derive(Debug, Default)]
+/// Moose is roaming the wilderness.
+struct Roam;
+
+async fn make_friend(
+    name: &str,
+    shared: &Arc<RwLock<SharedMooseState>>
+) -> Option<String> {
+    let mut mooses = shared.write().await;
+    for (other_moose, friends) in mooses.friends.iter_mut() {
+        if name == other_moose {
+            continue;
+        }
+        if !friends.contains(name) {
+            friends.insert(name.to_string());
+            return Some(other_moose.to_string());
+        }
+    }
+    return None;
+}
+
+#[async_trait::async_trait]
+impl State<MooseState> for Roam {
+    async fn next(
+        self: Box<Self>,
+        shared: Arc<RwLock<SharedMooseState>>,
+        state: &mut MooseState,
+        _manifest: Manifest<Moose>,
+    ) -> Transition<MooseState> {
+        loop {
+            tokio::time::delay_for(std::time::Duration::from_secs(5)).await;
+            state.food -= 5.0;
+            if state.food <= 50.0 {
+                return Transition::next(self, Eat);
+            }
+            let r: f64 = {
+                let mut rng = rand::thread_rng();
+                rng.gen()
+            };
+            if r < 0.05 {
+                if let Some(other_moose) = make_friend(&state.name, &shared).await {
+                    info!("{} made friends with {}!", state.name, other_moose);
+                }
+            }
+        }
+    }
+
+    async fn status(
+        &self,
+        _state: &mut MooseState,
+        _manifest: &Moose,
+    ) -> anyhow::Result<MooseStatus> {
+        Ok(MooseStatus {
+            phase: Some(MoosePhase::Roaming),
+            message: Some("Gahrooo!".to_string()),
+        })
+    }
+}
+```
+
+To finish our state machine, we must define valid transitions between states.
+The compiler will ensure that we do not make invalid transitions, which helps
+us to avoid undefined behavior. For Kubelet there exists a macro to derive this
```

Comment:
A link to the macro here would be great.

Reply:
Are you thinking a link to docs.rs? Could be tricky, since we do not have a release cut with this macro included. Also, the documentation for the marcro isn't especially verbose. Perhaps it would be better to include an example here?

Reply:
Not necessarily, maybe directly to the source code.